### PR TITLE
Implement Android emulator selection

### PR DESCRIPTION
**Pull request by JiglioNero**
#490

### CODE REVIEW

File: src/common/node/childProcess.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for details.
+
+import * as nodeChildProcess from "child_process";
+
+// Uncomment the following lines to record all spawned processes executions
+// import {Recorder} from "../../../test/resources/processExecution/recorder";
+// Recorder.installGlobalRecorder();
```
Comment (SounD120) [inline src/common/node/childProcess.ts:8]: Please remove not relevant comments

File: src/utils/android/adb.ts [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    ICE_CREAM_SANDWICH_MR1 = 15,
+    GINGERBREAD_MR1 = 10,
+}
+
+export enum DeviceType {
+    AndroidSdkEmulator, // These seem to have emulator-<port> ids
+    Other,
+}
+
+
```
Comment (SounD120) [inline src/utils/android/adb.ts:34]:
```suggestion
```

File: src/debugger/cordovaDebugSession.ts [hunk 3]
```diff
@@ -320,7 +320,7 @@ export class CordovaDebugSession extends LoggingDebugSession {
                         return this.cordovaCdpProxy.createServer(this.cdpProxyLogLevel, this.cancellationTokenSource.token);
                     })
                     .then(() => {
-                        if (target === TargetType.Device || target === TargetType.Emulator) {
+                        if ((platform === PlatformType.Android || platform === PlatformType.IOS) && (target !== TargetType.Edge && target !== TargetType.Chrome)) {
```
Comment (RedMickey) [inline src/debugger/cordovaDebugSession.ts:323]: I think it's better to check if the target is not a simulate target. We have the [`isSimulateTarget`](https://github.com/microsoft/vscode-cordova/blob/master/src/utils/simulateHelper.ts#L7) function.

File: src/utils/android/adb.ts [hunk 4]
```diff
@@ -72,6 +72,16 @@ export class AdbHelper {
         });
     }
 
+    public async getAvdNameById(emulatorId: string): Promise<string | null> {
+        return this.childProcess.execToString(`${this.adbExecutable} -s ${emulatorId} emu avd name`).then(output => {
+            if (output) {
+                return output.split(/\r?\n|\r/g)[0];
```
Comment (RedMickey) [inline src/utils/android/adb.ts:78]: Could you please add comments to show what the child process returns and what we fetch further?

File: src/debugger/cordovaDebugSession.ts [hunk 5]
```diff
@@ -1270,22 +1273,49 @@ To get the list of addresses run "ionic cordova run PLATFORM --livereload" (wher
...
+        targetArgs.push("--verbose");
+        if (!isDevice) {
+            const targetDevice = await androidEmulatorManager.startEmulator(launchArgs.target);
+            if (targetDevice) {
+                targetArgs.push("--emulator", `--target=${targetDevice.id}`);
+                if (isEmulator && targetDevice.name) {
+                    launchScenariousManager.updateLaunchScenario(launchArgs, {target: targetDevice.name});
+                }
+                launchArgs.target = targetDevice.id;
+            } else if (!launchArgs.target.toLowerCase().includes("emulator")) {
```
Comment (RedMickey) [inline src/debugger/cordovaDebugSession.ts:1295]:
```suggestion
            } else if (!launchArgs.target.toLowerCase().includes(TargetType.Emulator)) {
```

File: src/utils/android/androidEmulatorManager.ts [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                    emulatorName,
+                    AndroidEmulatorManager.EMULATOR_START_TIMEOUT,
+                )}`));
+            }, AndroidEmulatorManager.EMULATOR_START_TIMEOUT * 1000);
+
+            const bootCheckInterval = setInterval(async () => {
+                const connectedDevices = await this.adbHelper.getOnlineDevices();
+                for (let i = 0; i < connectedDevices.length; i++) {
+                    const onlineAvdName = await this.adbHelper.getAvdNameById(connectedDevices[i].id);
+                    console.log(onlineAvdName);
```
Comment (RedMickey) [inline src/utils/android/androidEmulatorManager.ts:108]:
```suggestion
```

**Review by RedMickey — APPROVED** [review]

**Pull request merged**